// @nota/nativescript-accessibility-ext/utils/AccessibilityHelper.android.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var nsApp = require("@nativescript/core/application");
var profiling_1 = require("@nativescript/core/profiling");
var trace = require("@nativescript/core/trace");
var view_1 = require("@nativescript/core/ui/core/view");
var gestures_1 = require("@nativescript/core/ui/gestures/gestures");
var list_view_1 = require("@nativescript/core/ui/list-view/list-view");
var proxy_view_container_1 = require("@nativescript/core/ui/proxy-view-container");
var utils = require("@nativescript/core/utils/utils");
var trace_1 = require("../trace");
var view_common_1 = require("../ui/core/view-common");
var helpers_1 = require("./helpers");
var utils_1 = require("./utils");
function writeHelperTrace(message, type) {
    if (type === void 0) { type = trace.messageType.info; }
    trace_1.writeTrace(message, type, trace_1.categories.AndroidHelper);
}
exports.getAndroidView = profiling_1.profile('getAndroidView', function getAndroidViewImpl(tnsView) {
    return tnsView.nativeView || tnsView.nativeViewProtected;
});
function getUIView(tnsView) {
    throw new Error("getUIView(" + tnsView + ") - should never be called on Android");
}
exports.getUIView = getUIView;
var AccessibilityEvent = android.view.accessibility.AccessibilityEvent;
var AccessibilityManager = android.view.accessibility.AccessibilityManager;
var AccessibilityDelegate = android.view.View.androidviewViewAccessibilityDelegate;
AccessibilityDelegate = android.view.View.AccessibilityDelegate;
var AccessibilityNodeInfo = android.view.accessibility.AccessibilityNodeInfo;
var AndroidView = android.view.View;
var AndroidViewGroup = android.view.ViewGroup;
var clickableRolesMap = new Set();
var getAccessibilityManager = profiling_1.profile('getAccessibilityManager', function getAccessibilityManagerImpl(view) {
    return view.getContext().getSystemService(android.content.Context.ACCESSIBILITY_SERVICE);
});
var suspendAccessibilityEvents = false;
var a11yScrollOnFocus = 'a11y-scroll-on-focus';
var lastFocusedView;
var accessibilityEventHelper = profiling_1.profile('accessibilityEventHelper', function accessibilityEventHelperImpl(tnsView, eventType) {
    var e_1, _a;
    var eventName = accessibilityEventTypeMap.get(eventType);
    if (!utils_1.isAccessibilityServiceEnabled()) {
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace("accessibilityEventHelper: Service not active");
        }
        return;
    }
    if (!eventName) {
        writeHelperTrace("accessibilityEventHelper: unknown eventType: " + eventType, trace.messageType.error);
        return;
    }
    if (!tnsView) {
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace("accessibilityEventHelper: no owner: " + eventName);
        }
        return;
    }
    var androidView = exports.getAndroidView(tnsView);
    if (!androidView) {
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace("accessibilityEventHelper: no nativeView");
        }
        return;
    }
    switch (eventType) {
        case AccessibilityEvent.TYPE_VIEW_CLICKED: {
            if (android.os.Build.VERSION.SDK_INT >= 26) {
                try {
                    for (var _b = __values(tnsView.getGestureObservers(gestures_1.GestureTypes.tap) || []), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var tapGesture = _c.value;
                        tapGesture.callback({
                            android: tnsView.android,
                            eventName: 'tap',
                            ios: null,
                            object: tnsView,
                            type: gestures_1.GestureTypes.tap,
                            view: tnsView,
                        });
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            return;
        }
        case AccessibilityEvent.TYPE_VIEW_ACCESSIBILITY_FOCUSED: {
            var lastView = lastFocusedView && lastFocusedView.get();
            if (lastView && tnsView !== lastView) {
                var lastAndroidView = exports.getAndroidView(lastView);
                if (lastAndroidView) {
                    lastAndroidView.clearFocus();
                    lastFocusedView = null;
                    helpers_1.notifyAccessibilityFocusState(lastView, false, true);
                }
            }
            lastFocusedView = new WeakRef(tnsView);
            helpers_1.notifyAccessibilityFocusState(tnsView, true, false);
            var tree = [];
            for (var node = tnsView; node; node = node.parent) {
                node.notify({
                    eventName: a11yScrollOnFocus,
                    object: tnsView,
                });
                tree.push(node + "[" + (node.className || '') + "]");
            }
            if (trace_1.isTraceEnabled()) {
                writeHelperTrace("Focus-tree: " + tree.reverse().join(' => '));
            }
            return;
        }
        case AccessibilityEvent.TYPE_VIEW_ACCESSIBILITY_FOCUS_CLEARED: {
            var lastView = lastFocusedView && lastFocusedView.get();
            if (lastView && tnsView === lastView) {
                lastFocusedView = null;
                androidView.clearFocus();
            }
            helpers_1.notifyAccessibilityFocusState(tnsView, false, true);
            return;
        }
    }
});
var TNSAccessibilityDelegate;
var androidViewToTNSView = new WeakMap();
var accessibilityEventMap;
var accessibilityEventTypeMap;
function ensureNativeClasses() {
    if (TNSAccessibilityDelegate) {
        return;
    }
    var RoleTypeMap = new Map([
        [view_common_1.AccessibilityRole.Button, android.widget.Button.class.getName()],
        [view_common_1.AccessibilityRole.Search, android.widget.EditText.class.getName()],
        [view_common_1.AccessibilityRole.Image, android.widget.ImageView.class.getName()],
        [view_common_1.AccessibilityRole.ImageButton, android.widget.ImageButton.class.getName()],
        [view_common_1.AccessibilityRole.KeyboardKey, android.inputmethodservice.Keyboard.Key.class.getName()],
        [view_common_1.AccessibilityRole.StaticText, android.widget.TextView.class.getName()],
        [view_common_1.AccessibilityRole.Adjustable, android.widget.SeekBar.class.getName()],
        [view_common_1.AccessibilityRole.Checkbox, android.widget.CheckBox.class.getName()],
        [view_common_1.AccessibilityRole.RadioButton, android.widget.RadioButton.class.getName()],
        [view_common_1.AccessibilityRole.SpinButton, android.widget.Spinner.class.getName()],
        [view_common_1.AccessibilityRole.Switch, android.widget.Switch.class.getName()],
        [view_common_1.AccessibilityRole.ProgressBar, android.widget.ProgressBar.class.getName()],
    ]);
    clickableRolesMap = new Set([view_common_1.AccessibilityRole.Button, view_common_1.AccessibilityRole.ImageButton]);
    var ignoreRoleTypesForTrace = new Set([view_common_1.AccessibilityRole.Header, view_common_1.AccessibilityRole.Link, view_common_1.AccessibilityRole.None, view_common_1.AccessibilityRole.Summary]);
    var TNSAccessibilityDelegateImpl = (function (_super) {
        __extends(TNSAccessibilityDelegateImpl, _super);
        function TNSAccessibilityDelegateImpl() {
            var _this = _super.call(this) || this;
            return global.__native(_this);
        }
        TNSAccessibilityDelegateImpl.prototype.getTnsView = function (view) {
            if (!androidViewToTNSView.has(view)) {
                return null;
            }
            var tnsView = androidViewToTNSView.get(view).get();
            if (!tnsView) {
                androidViewToTNSView.delete(view);
                return null;
            }
            return tnsView;
        };
        TNSAccessibilityDelegateImpl.prototype.onInitializeAccessibilityNodeInfo = function (host, info) {
            _super.prototype.onInitializeAccessibilityNodeInfo.call(this, host, info);
            var tnsView = this.getTnsView(host);
            if (!tnsView) {
                if (trace_1.isTraceEnabled()) {
                    writeHelperTrace("onInitializeAccessibilityNodeInfo " + host + " " + info + " no tns-view");
                }
                return;
            }
            var accessibilityRole = tnsView.accessibilityRole;
            if (accessibilityRole) {
                var androidClassName = RoleTypeMap.get(accessibilityRole);
                if (androidClassName) {
                    var oldClassName = info.getClassName() || (android.os.Build.VERSION.SDK_INT >= 28 && host.getAccessibilityClassName()) || null;
                    info.setClassName(androidClassName);
                    if (trace_1.isTraceEnabled()) {
                        writeHelperTrace(tnsView + ".accessibilityRole = \"" + accessibilityRole + "\" is mapped to \"" + androidClassName + "\" (was " + oldClassName + "). " + info.getClassName());
                    }
                }
                else if (!ignoreRoleTypesForTrace.has(accessibilityRole)) {
                    if (trace_1.isTraceEnabled()) {
                        writeHelperTrace(tnsView + ".accessibilityRole = \"" + accessibilityRole + "\" is unknown");
                    }
                }
                if (clickableRolesMap.has(accessibilityRole)) {
                    if (trace_1.isTraceEnabled()) {
                        writeHelperTrace("onInitializeAccessibilityNodeInfo " + tnsView + " - set clickable role=" + accessibilityRole);
                    }
                    info.setClickable(true);
                }
                if (android.os.Build.VERSION.SDK_INT >= 28) {
                    if (accessibilityRole === view_common_1.AccessibilityRole.Header) {
                        if (trace_1.isTraceEnabled()) {
                            writeHelperTrace("onInitializeAccessibilityNodeInfo " + tnsView + " - set heading role=" + accessibilityRole);
                        }
                        info.setHeading(true);
                    }
                    else if (host.isAccessibilityHeading()) {
                        if (trace_1.isTraceEnabled()) {
                            writeHelperTrace("onInitializeAccessibilityNodeInfo " + tnsView + " - set heading from host");
                        }
                        info.setHeading(true);
                    }
                    else {
                        if (trace_1.isTraceEnabled()) {
                            writeHelperTrace("onInitializeAccessibilityNodeInfo " + tnsView + " - set not heading");
                        }
                        info.setHeading(false);
                    }
                }
                switch (accessibilityRole) {
                    case view_common_1.AccessibilityRole.Switch:
                    case view_common_1.AccessibilityRole.RadioButton:
                    case view_common_1.AccessibilityRole.Checkbox: {
                        if (trace_1.isTraceEnabled()) {
                            writeHelperTrace("onInitializeAccessibilityNodeInfo " + tnsView + " - set checkable and check=" + (tnsView.accessibilityState === view_common_1.AccessibilityState.Checked));
                        }
                        info.setCheckable(true);
                        info.setChecked(tnsView.accessibilityState === view_common_1.AccessibilityState.Checked);
                        break;
                    }
                    default: {
                        if (trace_1.isTraceEnabled()) {
                            writeHelperTrace("onInitializeAccessibilityNodeInfo " + tnsView + " - set enabled=" + (tnsView.accessibilityState !==
                                view_common_1.AccessibilityState.Disabled) + " and selected=" + (tnsView.accessibilityState === view_common_1.AccessibilityState.Selected));
                        }
                        info.setEnabled(tnsView.accessibilityState !== view_common_1.AccessibilityState.Disabled);
                        info.setSelected(tnsView.accessibilityState === view_common_1.AccessibilityState.Selected);
                        break;
                    }
                }
            }
            if (tnsView.accessible === true) {
                info.setFocusable(true);
            }
        };
        TNSAccessibilityDelegateImpl.prototype.onInitializeAccessibilityEvent = function (view, event) {
            _super.prototype.onInitializeAccessibilityEvent.call(this, view, event);
        };
        TNSAccessibilityDelegateImpl.prototype.onPopulateAccessibilityEvent = function (view, event) {
            _super.prototype.onPopulateAccessibilityEvent.call(this, view, event);
        };
        TNSAccessibilityDelegateImpl.prototype.dispatchPopulateAccessibilityEvent = function (view, event) {
            return _super.prototype.dispatchPopulateAccessibilityEvent.call(this, view, event);
        };
        TNSAccessibilityDelegateImpl.prototype.sendAccessibilityEvent = function (host, eventType) {
            _super.prototype.sendAccessibilityEvent.call(this, host, eventType);
            var tnsView = this.getTnsView(host);
            if (!tnsView) {
                console.log("skip - " + host + " - " + accessibilityEventTypeMap.get(eventType));
                return;
            }
            if (suspendAccessibilityEvents) {
                if (trace_1.isTraceEnabled()) {
                    writeHelperTrace("sendAccessibilityEvent: " + tnsView + " - skip");
                }
                return;
            }
            try {
                accessibilityEventHelper(tnsView, eventType);
            }
            catch (err) {
                console.error(err);
            }
        };
        TNSAccessibilityDelegateImpl.prototype.onRequestSendAccessibilityEvent = function (host, view, event) {
            if (suspendAccessibilityEvents) {
                return false;
            }
            return _super.prototype.onRequestSendAccessibilityEvent.call(this, host, view, event);
        };
        return TNSAccessibilityDelegateImpl;
    }(AccessibilityDelegate));
    TNSAccessibilityDelegate = new TNSAccessibilityDelegateImpl();
    accessibilityEventMap = new Map([
        ['invalid_position', AccessibilityEvent.INVALID_POSITION],
        ['max_text_length', AccessibilityEvent.MAX_TEXT_LENGTH],
        ['view_clicked', AccessibilityEvent.TYPE_VIEW_CLICKED],
        ['view_long_clicked', AccessibilityEvent.TYPE_VIEW_LONG_CLICKED],
        ['view_selected', AccessibilityEvent.TYPE_VIEW_SELECTED],
        ['view_focused', AccessibilityEvent.TYPE_VIEW_FOCUSED],
        ['view_text_changed', AccessibilityEvent.TYPE_VIEW_TEXT_CHANGED],
        ['window_state_changed', AccessibilityEvent.TYPE_WINDOW_STATE_CHANGED],
        ['notification_state_changed', AccessibilityEvent.TYPE_NOTIFICATION_STATE_CHANGED],
        ['view_hover_enter', AccessibilityEvent.TYPE_VIEW_HOVER_ENTER],
        ['view_hover_exit', AccessibilityEvent.TYPE_VIEW_HOVER_EXIT],
        ['touch_exploration_gesture_start', AccessibilityEvent.TYPE_TOUCH_EXPLORATION_GESTURE_START],
        ['touch_exploration_gesture_end', AccessibilityEvent.TYPE_TOUCH_EXPLORATION_GESTURE_END],
        ['window_content_changed', AccessibilityEvent.TYPE_WINDOW_CONTENT_CHANGED],
        ['view_scrolled', AccessibilityEvent.TYPE_VIEW_SCROLLED],
        ['view_text_selection_changed', AccessibilityEvent.TYPE_VIEW_TEXT_SELECTION_CHANGED],
        ['announcement', AccessibilityEvent.TYPE_ANNOUNCEMENT],
        ['view_accessibility_focused', AccessibilityEvent.TYPE_VIEW_ACCESSIBILITY_FOCUSED],
        ['view_accessibility_focus_cleared', AccessibilityEvent.TYPE_VIEW_ACCESSIBILITY_FOCUS_CLEARED],
        ['view_text_traversed_at_movement_granularity', AccessibilityEvent.TYPE_VIEW_TEXT_TRAVERSED_AT_MOVEMENT_GRANULARITY],
        ['gesture_detection_start', AccessibilityEvent.TYPE_GESTURE_DETECTION_START],
        ['gesture_detection_end', AccessibilityEvent.TYPE_GESTURE_DETECTION_END],
        ['touch_interaction_start', AccessibilityEvent.TYPE_TOUCH_INTERACTION_START],
        ['touch_interaction_end', AccessibilityEvent.TYPE_TOUCH_INTERACTION_END],
        ['all', AccessibilityEvent.TYPES_ALL_MASK],
    ]);
    accessibilityEventTypeMap = new Map(__spread(accessibilityEventMap).map(function (_a) {
        var _b = __read(_a, 2), k = _b[0], v = _b[1];
        return [v, k];
    }));
}
var AccessibilityHelper = (function () {
    function AccessibilityHelper() {
    }
    AccessibilityHelper.updateAccessibilityProperties = function (tnsView) {
        if (tnsView instanceof proxy_view_container_1.ProxyViewContainer) {
            return null;
        }
        setAccessibilityDelegate(tnsView);
        applyContentDescription(tnsView);
    };
    AccessibilityHelper.sendAccessibilityEvent = function (tnsView, eventName, text) {
        var cls = "AccessibilityHelper.sendAccessibilityEvent(" + tnsView + ", " + eventName + ", " + text + ")";
        var androidView = exports.getAndroidView(tnsView);
        if (!androidView) {
            if (trace_1.isTraceEnabled()) {
                writeHelperTrace(cls + ": no nativeView");
            }
            return;
        }
        if (!eventName) {
            if (trace_1.isTraceEnabled()) {
                writeHelperTrace(cls + ": no eventName provided");
            }
            return;
        }
        if (!utils_1.isAccessibilityServiceEnabled()) {
            if (trace_1.isTraceEnabled()) {
                writeHelperTrace(cls + " - TalkBack not enabled");
            }
            return;
        }
        var a11yService = getAccessibilityManager(androidView);
        if (!a11yService.isEnabled()) {
            if (trace_1.isTraceEnabled()) {
                writeHelperTrace(cls + " - a11yService not enabled");
            }
            return;
        }
        eventName = eventName.toLowerCase();
        if (!accessibilityEventMap.has(eventName)) {
            if (trace_1.isTraceEnabled()) {
                writeHelperTrace(cls + " - unknown event");
            }
            return;
        }
        var eventInt = accessibilityEventMap.get(eventName);
        if (!text) {
            return androidView.sendAccessibilityEvent(eventInt);
        }
        var a11yEvent = AccessibilityEvent.obtain(eventInt);
        a11yEvent.setSource(androidView);
        a11yEvent.getText().clear();
        if (!text) {
            applyContentDescription(tnsView);
            text = androidView.getContentDescription() || tnsView['title'];
            if (trace_1.isTraceEnabled()) {
                writeHelperTrace(cls + " - text not provided use androidView.getContentDescription() - " + text);
            }
        }
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace(cls + ": send event with text: '" + JSON.stringify(text) + "'");
        }
        if (text) {
            a11yEvent.getText().add(text);
        }
        a11yService.sendAccessibilityEvent(a11yEvent);
    };
    AccessibilityHelper.updateContentDescription = function (tnsView, forceUpdate) {
        if (forceUpdate === void 0) { forceUpdate = false; }
        if (tnsView instanceof proxy_view_container_1.ProxyViewContainer) {
            return null;
        }
        return applyContentDescription(tnsView, forceUpdate);
    };
    __decorate([
        profiling_1.profile,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [view_1.View]),
        __metadata("design:returntype", void 0)
    ], AccessibilityHelper, "updateAccessibilityProperties", null);
    __decorate([
        profiling_1.profile,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [view_1.View, String, String]),
        __metadata("design:returntype", void 0)
    ], AccessibilityHelper, "sendAccessibilityEvent", null);
    __decorate([
        profiling_1.profile,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [view_1.View, Object]),
        __metadata("design:returntype", void 0)
    ], AccessibilityHelper, "updateContentDescription", null);
    return AccessibilityHelper;
}());
exports.AccessibilityHelper = AccessibilityHelper;
var removeAccessibilityDelegate = profiling_1.profile('removeAccessibilityDelegate', function removeAccessibilityDelegateImpl(tnsView) {
    if (tnsView instanceof proxy_view_container_1.ProxyViewContainer) {
        return null;
    }
    var androidView = exports.getAndroidView(tnsView);
    if (!androidView) {
        return;
    }
    androidViewToTNSView.delete(androidView);
    androidView.setAccessibilityDelegate(null);
});
var setAccessibilityDelegate = profiling_1.profile('setAccessibilityDelegate', function setAccessibilityDelegateImpl(tnsView) {
    if (tnsView instanceof proxy_view_container_1.ProxyViewContainer) {
        return null;
    }
    ensureNativeClasses();
    var androidView = exports.getAndroidView(tnsView);
    if (!androidView) {
        return;
    }
    androidViewToTNSView.set(androidView, new WeakRef(tnsView));
    var hasOldDelegate = androidView.getAccessibilityDelegate() === TNSAccessibilityDelegate;
    var cls = "AccessibilityHelper.updateAccessibilityProperties(" + tnsView + ") - has delegate? " + hasOldDelegate;
    if (trace_1.isTraceEnabled()) {
        writeHelperTrace(cls);
    }
    if (hasOldDelegate) {
        return;
    }
    androidView.setAccessibilityDelegate(TNSAccessibilityDelegate);
});
var applyContentDescription = profiling_1.profile('applyContentDescription', function applyContentDescriptionImpl(tnsView, forceUpdate) {
    if (tnsView instanceof proxy_view_container_1.ProxyViewContainer) {
        return null;
    }
    var androidView = exports.getAndroidView(tnsView);
    if (androidView instanceof androidx.appcompat.widget.Toolbar) {
        var numChildren = androidView.getChildCount();
        for (var i = 0; i < numChildren; i += 1) {
            var childAndroidView = androidView.getChildAt(i);
            if (childAndroidView instanceof androidx.appcompat.widget.AppCompatTextView) {
                androidView = childAndroidView;
                break;
            }
        }
    }
    var cls = "applyContentDescription(" + tnsView + ")";
    if (!androidView) {
        return null;
    }
    var titleValue = (tnsView['title'] || '').toString().trim();
    var textValue = (tnsView['text'] || '').toString().trim();
    if (!forceUpdate && tnsView._androidContentDescriptionUpdated === false && textValue === tnsView['_lastText'] && titleValue === tnsView['_lastTitle']) {
        return androidView.getContentDescription();
    }
    var contentDescriptionBuilder = new Array();
    if (tnsView.accessibilityRole === view_common_1.AccessibilityRole.Switch) {
        var androidSwitch = new android.widget.Switch(nsApp.android.context);
        if (tnsView.accessibilityState === view_common_1.AccessibilityState.Checked) {
            contentDescriptionBuilder.push(androidSwitch.getTextOn());
        }
        else {
            contentDescriptionBuilder.push(androidSwitch.getTextOff());
        }
    }
    var accessibilityLabel = (tnsView.accessibilityLabel || '').trim();
    var accessibilityValue = (tnsView.accessibilityValue || '').trim();
    var accessibilityHint = (tnsView.accessibilityHint || '').trim();
    if (accessibilityLabel) {
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace(cls + " - have accessibilityLabel");
        }
        contentDescriptionBuilder.push("" + accessibilityLabel);
    }
    if (accessibilityValue) {
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace(cls + " - have accessibilityValue");
        }
        contentDescriptionBuilder.push("" + accessibilityValue);
    }
    else if (textValue) {
        if (!accessibilityLabel || textValue.toLowerCase() !== accessibilityLabel.toLowerCase()) {
            if (trace_1.isTraceEnabled()) {
                writeHelperTrace(cls + " - don't have accessibilityValue - use 'text' value");
            }
            contentDescriptionBuilder.push("" + textValue);
        }
    }
    else if (titleValue) {
        if (!accessibilityLabel || titleValue.toLowerCase() !== accessibilityLabel.toLowerCase()) {
            if (trace_1.isTraceEnabled()) {
                writeHelperTrace(cls + " - don't have accessibilityValue - use 'title' value");
            }
            contentDescriptionBuilder.push("" + titleValue);
        }
    }
    if (accessibilityHint) {
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace(cls + " - have accessibilityHint");
        }
        contentDescriptionBuilder.push("" + accessibilityHint);
    }
    var contentDescription = contentDescriptionBuilder
        .join('. ')
        .trim()
        .replace(/^\.$/, '');
    if (contentDescription) {
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace(cls + " - set to \"" + contentDescription + "\"");
        }
        androidView.setContentDescription(contentDescription);
    }
    else {
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace(cls + " - remove value");
        }
        androidView.setContentDescription(null);
    }
    tnsView['_lastTitle'] = titleValue;
    tnsView['_lastText'] = textValue;
    tnsView._androidContentDescriptionUpdated = false;
    return contentDescription;
});
var ensureListViewItemIsOnScreen = profiling_1.profile('ensureListViewItemIsOnScreen', function ensureListViewItemIsOnScreenImpl(listView, tnsView) {
    if (suspendAccessibilityEvents) {
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace("ensureListViewItemIsOnScreen(" + listView + ", " + tnsView + ") suspended");
        }
        return;
    }
    if (trace_1.isTraceEnabled()) {
        writeHelperTrace("ensureListViewItemIsOnScreen(" + listView + ", " + tnsView + ")");
    }
    var androidListView = exports.getAndroidView(listView);
    if (!androidListView) {
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace("ensureListViewItemIsOnScreen(" + listView + ", " + tnsView + ") no native list-view?");
        }
        return;
    }
    var androidView = exports.getAndroidView(tnsView);
    if (!androidView) {
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace("ensureListViewItemIsOnScreen(" + listView + ", " + tnsView + ") no native item view?");
        }
        return;
    }
    try {
        removeAccessibilityDelegate(tnsView);
        suspendAccessibilityEvents = true;
        var viewSize = tnsView.getActualSize();
        var viewPos = tnsView.getLocationRelativeTo(listView);
        var listViewSize = listView.getActualSize();
        var viewPosDelta = {
            x2: viewSize.width + viewPos.x,
            y2: viewSize.height + viewPos.y,
        };
        var offsetPadding = 10;
        var minOffset = offsetPadding;
        var maxOffset = listViewSize.height - offsetPadding;
        if (viewPos.y >= minOffset && viewPosDelta.y2 <= maxOffset) {
            if (trace_1.isTraceEnabled()) {
                writeHelperTrace("ensureListViewItemIsOnScreen(" + listView + ", " + tnsView + ") view is on screen " + viewPos.y + " >= " + minOffset + " && " + viewPosDelta.y2 + " <= " + maxOffset);
            }
            return;
        }
        var wantedScrollOffset = viewPos.y < 0 ? offsetPadding : listViewSize.height - viewSize.height - offsetPadding;
        var scrollByDIP = viewPos.y - wantedScrollOffset;
        var scrollByDP = utils.layout.toDevicePixels(scrollByDIP);
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace("ensureListViewItemIsOnScreen(" + listView + ", " + tnsView + ") view is not on screen, scroll by: " + scrollByDIP);
        }
        androidx.core.widget.ListViewCompat.scrollListBy(androidListView, scrollByDP);
    }
    catch (err) {
        trace_1.writeErrorTrace(err);
    }
    finally {
        suspendAccessibilityEvents = false;
        AccessibilityHelper.updateAccessibilityProperties(tnsView);
    }
});
function setupA11yScrollOnFocus(args) {
    var listView = args.object;
    var tnsView = args.view;
    if (!tnsView) {
        return;
    }
    setAccessibilityDelegate(tnsView);
    if (tnsView.hasListeners(a11yScrollOnFocus)) {
        if (trace_1.isTraceEnabled()) {
            writeHelperTrace("setupA11yScrollOnFocus(): " + listView + " view=" + tnsView + " - item already has " + a11yScrollOnFocus);
        }
        return;
    }
    if (trace_1.isTraceEnabled()) {
        writeHelperTrace("setupA11yScrollOnFocus(): " + listView + " view=" + tnsView);
    }
    var listViewRef = new WeakRef(listView);
    tnsView.on(a11yScrollOnFocus, function (evt) {
        var localListView = listViewRef.get();
        if (!localListView) {
            evt.object.off(a11yScrollOnFocus);
            return;
        }
        ensureListViewItemIsOnScreen(localListView, evt.object);
    });
}
helpers_1.hmrSafeEvents('setupA11yScrollOnFocus', [list_view_1.ListView.itemLoadingEvent], list_view_1.ListView, setupA11yScrollOnFocus);
helpers_1.hmrSafeEvents('setAccessibilityDelegate:loadedEvent', [view_1.View.loadedEvent], view_1.View, function (evt) {
    AccessibilityHelper.updateAccessibilityProperties(evt.object);
});
