// @nota/nativescript-accessibility-ext/angular/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
require("../index");
var a11y_font_scaling_1 = require("./data/a11y-font-scaling");
exports.A11yFontScalingObservable = a11y_font_scaling_1.A11yFontScalingObservable;
var a11y_service_enabled_1 = require("./data/a11y-service-enabled");
exports.A11yServiceEnabledObservable = a11y_service_enabled_1.A11yServiceEnabledObservable;
var a11y_grid_layout_directive_1 = require("./directives/a11y-grid-layout.directive");
var a11y_font_scaling_pipe_1 = require("./pipes/a11y-font-scaling.pipe");
exports.A11YFontScalePipe = a11y_font_scaling_pipe_1.A11YFontScalePipe;
var NotaAccessibilityExtModule = (function () {
    function NotaAccessibilityExtModule() {
    }
    NotaAccessibilityExtModule = __decorate([
        core_1.NgModule({
            providers: [a11y_font_scaling_pipe_1.A11YFontScalePipe],
            declarations: [a11y_grid_layout_directive_1.A11YGridLayoutDirective, a11y_font_scaling_pipe_1.A11YFontScalePipe],
            exports: [a11y_grid_layout_directive_1.A11YGridLayoutDirective, a11y_font_scaling_pipe_1.A11YFontScalePipe],
        })
    ], NotaAccessibilityExtModule);
    return NotaAccessibilityExtModule;
}());
exports.NotaAccessibilityExtModule = NotaAccessibilityExtModule;
