// @nota/nativescript-accessibility-ext/angular/services/base.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var rxjs_1 = require("rxjs");
var BaseService = (function () {
    function BaseService() {
        this.isDeleted = false;
        this.destroy$ = new rxjs_1.ReplaySubject(1);
    }
    BaseService.prototype.ngOnDestroy = function () {
        var e_1, _a;
        this.destroy$.next(true);
        this.isDeleted = true;
        try {
            for (var _b = __values(Object.entries(this)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), key = _d[0], value = _d[1];
                if (!value) {
                    continue;
                }
                if (value instanceof core_1.ElementRef) {
                    delete this[key];
                    continue;
                }
                if (value instanceof rxjs_1.Subscription) {
                    try {
                        value.unsubscribe();
                    }
                    catch (_e) {
                    }
                    delete this[key];
                    continue;
                }
                if (value instanceof rxjs_1.Observable) {
                    if (key === 'destroy$') {
                        continue;
                    }
                    delete this[key];
                    continue;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    BaseService.prototype.takeUntilDestroy = function () {
        var destroy$ = this.destroy$;
        return function (source) {
            return new rxjs_1.Observable(function (subscriber) {
                var sub = source.subscribe(subscriber);
                var destroySub = destroy$.subscribe(function () {
                    if (sub) {
                        sub.unsubscribe();
                    }
                    if (subscriber) {
                        subscriber.complete();
                    }
                    if (destroySub) {
                        destroySub.unsubscribe();
                    }
                });
                return function () {
                    sub.unsubscribe();
                    destroySub.unsubscribe();
                };
            });
        };
    };
    BaseService.prototype.timerUnlessDestroyed = function (cb, delay) {
        if (delay === void 0) { delay = 0; }
        return rxjs_1.timer(delay)
            .pipe(this.takeUntilDestroy())
            .subscribe(function () {
            try {
                cb();
            }
            catch (_a) {
            }
        });
    };
    return BaseService;
}());
exports.BaseService = BaseService;
