// @nota/nativescript-accessibility-ext/trace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var trace = require("@nativescript/core/trace");
var categories;
(function (categories) {
    categories.A11Y = 'A11Y';
    categories.GlobalEvents = categories.A11Y + "-GlobalEvents";
    categories.FontScale = categories.A11Y + "-FontScale";
    categories.AndroidHelper = categories.A11Y + "-AndroidHelper";
    categories.separator = ',';
    categories.All = [categories.A11Y, categories.GlobalEvents, categories.FontScale, categories.AndroidHelper];
})(categories = exports.categories || (exports.categories = {}));
function isTraceEnabled() {
    return trace.isEnabled();
}
exports.isTraceEnabled = isTraceEnabled;
function writeTrace(message, type, category) {
    if (type === void 0) { type = trace.messageType.info; }
    if (category === void 0) { category = categories.A11Y; }
    if (isTraceEnabled()) {
        trace.write(message, category, type);
    }
}
exports.writeTrace = writeTrace;
function writeFontScaleTrace(message, type) {
    if (type === void 0) { type = trace.messageType.info; }
    writeTrace(message, type, categories.FontScale);
}
exports.writeFontScaleTrace = writeFontScaleTrace;
function writeGlobalEventsTrace(message, type) {
    if (type === void 0) { type = trace.messageType.info; }
    writeTrace(message, type, categories.GlobalEvents);
}
exports.writeGlobalEventsTrace = writeGlobalEventsTrace;
function writeErrorTrace(message) {
    trace.write(message, categories.A11Y, trace.messageType.error);
}
exports.writeErrorTrace = writeErrorTrace;
function writeWarnTrace(message) {
    writeTrace(message, trace.messageType.warn);
}
exports.writeWarnTrace = writeWarnTrace;
