// @nota/nativescript-accessibility-ext/ui/css-classes-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var nsApp = require("@nativescript/core/application/application");
var system_classes_1 = require("@nativescript/core/css/system-classes");
var platform_1 = require("@nativescript/core/platform");
var profiling_1 = require("@nativescript/core/profiling");
var view_1 = require("@nativescript/core/ui/core/view");
var trace_1 = require("../trace");
var FontScaleObservable_1 = require("../utils/FontScaleObservable");
require("../utils/global-events");
var helpers_1 = require("../utils/helpers");
var utils_1 = require("../utils/utils");
var view_common_1 = require("./core/view-common");
var fontExtraSmallClass = "a11y-fontscale-xs";
var fontExtraMediumClass = "a11y-fontscale-m";
var fontExtraLargeClass = "a11y-fontscale-xl";
var a11yServiceEnabledClass = "a11y-service-enabled";
var a11yServiceDisabledClass = "a11y-service-disabled";
var rootA11YClass = 'ns-a11y';
var A11YCssClassHelper = (function () {
    function A11YCssClassHelper() {
        var _this = this;
        this.cls = "CssClassesHelper";
        this.fontScaleCssClasses = new Map(FontScaleObservable_1.FontScaleObservable.VALID_FONT_SCALES.map(function (fs) { return [fs, "a11y-fontscale-" + Number(fs * 100).toFixed(0)]; }));
        this.currentFontScaleClass = '';
        this.currentFontScaleCategory = '';
        this.currentA11YStatusClass = '';
        this.currentRootA11YClass = '';
        this.fontScaleObservable = new FontScaleObservable_1.FontScaleObservable();
        this.a11yServiceObservable = new utils_1.AccessibilityServiceEnabledObservable();
        this.loadedModalViewRefs = new Map();
        this.fontScaleObservable.on(FontScaleObservable_1.FontScaleObservable.propertyChangeEvent, this.fontScaleChanged, this);
        this.a11yServiceObservable.on(utils_1.AccessibilityServiceEnabledObservable.propertyChangeEvent, this.a11yServiceChanged, this);
        helpers_1.hmrSafeEvents(this.cls + ".updateRootViews", [nsApp.displayedEvent, nsApp.resumeEvent], nsApp, function (evt) {
            _this.updateRootViews(evt);
        });
        helpers_1.hmrSafeEvents(this.cls + ".modalViewShowing", [view_1.View.shownModallyEvent], view_common_1.ViewCommon, function (evt) {
            _this.addModalViewRef(evt.object);
        });
    }
    A11YCssClassHelper.prototype.updateRootViews = function (evt) {
        var e_1, _a;
        evt = __assign({}, evt);
        var cls = this.cls + ".updateRootViews({eventName: " + evt.eventName + ", object: " + evt.object + "})";
        if (!this.updateCurrentHelperClasses()) {
            if (trace_1.isTraceEnabled()) {
                trace_1.writeTrace(cls + " - no changes");
            }
            return;
        }
        try {
            for (var _b = __values(__spread([nsApp.getRootView()], this.getModalViews())), _c = _b.next(); !_c.done; _c = _b.next()) {
                var view = _c.value;
                if (!view) {
                    continue;
                }
                if (trace_1.isTraceEnabled()) {
                    trace_1.writeTrace(cls + " - update css state on " + view);
                }
                view._onCssStateChange();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    A11YCssClassHelper.prototype.getModalViews = function () {
        var e_2, _a;
        var views = [];
        try {
            for (var _b = __values(this.loadedModalViewRefs), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), id = _d[0], viewRef = _d[1];
                var view = viewRef.get();
                if (!view) {
                    this.loadedModalViewRefs.delete(id);
                    continue;
                }
                views.push(view);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return views;
    };
    A11YCssClassHelper.prototype.addModalViewRef = function (modalView) {
        var e_3, _a;
        try {
            for (var _b = __values(this.loadedModalViewRefs), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), id = _d[0], viewRef = _d[1];
                var otherView = viewRef.get();
                if (!otherView) {
                    this.loadedModalViewRefs.delete(id);
                    continue;
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
        this.loadedModalViewRefs.set("" + modalView, new WeakRef(modalView));
    };
    A11YCssClassHelper.prototype.removeCssClass = function (cssClass) {
        system_classes_1.removeSystemCssClass(cssClass);
        __spread([nsApp.getRootView()], this.getModalViews()).forEach(function (view) {
            if (view) {
                view.cssClasses.delete(cssClass);
            }
        });
    };
    A11YCssClassHelper.prototype.addCssClass = function (cssClass) {
        system_classes_1.pushToRootViewCssClasses(cssClass);
        __spread([nsApp.getRootView()], this.getModalViews()).forEach(function (view) {
            if (view) {
                view.cssClasses.add(cssClass);
            }
        });
    };
    A11YCssClassHelper.prototype.updateCurrentHelperClasses = function () {
        var _a = this.fontScaleObservable, fontScale = _a.fontScale, isExtraSmall = _a.isExtraSmall, isExtraLarge = _a.isExtraLarge;
        var changed = false;
        var oldFontScaleClass = this.currentFontScaleClass;
        if (this.fontScaleCssClasses.has(fontScale)) {
            this.currentFontScaleClass = this.fontScaleCssClasses.get(fontScale);
        }
        else {
            this.currentFontScaleClass = this.fontScaleCssClasses.get(1);
        }
        if (oldFontScaleClass !== this.currentFontScaleClass) {
            this.removeCssClass(oldFontScaleClass);
            this.addCssClass(this.currentFontScaleClass);
            changed = true;
        }
        var oldActiveFontScaleCategory = this.currentFontScaleCategory;
        if (platform_1.isAndroid) {
            this.currentFontScaleCategory = fontExtraMediumClass;
        }
        else {
            if (isExtraSmall) {
                this.currentFontScaleCategory = fontExtraSmallClass;
            }
            else if (isExtraLarge) {
                this.currentFontScaleCategory = fontExtraLargeClass;
            }
            else {
                this.currentFontScaleCategory = fontExtraMediumClass;
            }
        }
        if (oldActiveFontScaleCategory !== this.currentFontScaleCategory) {
            this.removeCssClass(oldActiveFontScaleCategory);
            this.addCssClass(this.currentFontScaleCategory);
            changed = true;
        }
        var oldA11YStatusClass = this.currentA11YStatusClass;
        if (this.a11yServiceObservable.accessibilityServiceEnabled) {
            this.currentA11YStatusClass = a11yServiceEnabledClass;
        }
        else {
            this.currentA11YStatusClass = a11yServiceDisabledClass;
        }
        if (oldA11YStatusClass !== this.currentA11YStatusClass) {
            this.removeCssClass(oldA11YStatusClass);
            this.addCssClass(this.currentA11YStatusClass);
            changed = true;
        }
        if (this.currentRootA11YClass !== rootA11YClass) {
            this.addCssClass(rootA11YClass);
            this.currentRootA11YClass = rootA11YClass;
            changed = true;
        }
        return changed;
    };
    A11YCssClassHelper.prototype.fontScaleChanged = function (event) {
        var fontScale = this.fontScaleObservable.fontScale;
        if (trace_1.isTraceEnabled()) {
            trace_1.writeFontScaleTrace(this.cls + ".fontScaleChanged(): " + FontScaleObservable_1.FontScaleObservable.FONT_SCALE + " changed to " + fontScale);
        }
        this.updateRootViews(event);
    };
    A11YCssClassHelper.prototype.a11yServiceChanged = function (event) {
        if (trace_1.isTraceEnabled()) {
            trace_1.writeFontScaleTrace(this.cls + ".a11yServiceChanged(): to " + this.a11yServiceObservable.accessibilityServiceEnabled);
        }
        this.updateRootViews(event);
    };
    __decorate([
        profiling_1.profile,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], A11YCssClassHelper.prototype, "updateRootViews", null);
    __decorate([
        profiling_1.profile,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], A11YCssClassHelper.prototype, "getModalViews", null);
    __decorate([
        profiling_1.profile,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [view_1.View]),
        __metadata("design:returntype", void 0)
    ], A11YCssClassHelper.prototype, "addModalViewRef", null);
    __decorate([
        profiling_1.profile,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], A11YCssClassHelper.prototype, "removeCssClass", null);
    __decorate([
        profiling_1.profile,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], A11YCssClassHelper.prototype, "addCssClass", null);
    __decorate([
        profiling_1.profile,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Boolean)
    ], A11YCssClassHelper.prototype, "updateCurrentHelperClasses", null);
    __decorate([
        profiling_1.profile,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], A11YCssClassHelper.prototype, "fontScaleChanged", null);
    __decorate([
        profiling_1.profile,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], A11YCssClassHelper.prototype, "a11yServiceChanged", null);
    return A11YCssClassHelper;
}());
exports.cssClassHelper = new A11YCssClassHelper();
